odoo/addons/base/models/ir_actions_report.py

```diff
@@ -410,6 +413,58 @@ def _prepare_html(self, html, report_model=False):
 
         return bodies, res_ids, header, footer, specific_paperformat_args
 
+    @api.model
+    def _build_wkhtmltoimage_args(self, width, height, allow_js=False):
+        '''Build arguments understandable by wkhtmltoimage bin.'''
+        command_args = ['--disable-local-file-access']
+
+        # Passing the cookie to wkhtmltopdf in order to resolve internal links.
+        if request and request.db:
+            command_args.extend(['--cookie', 'session_id', request.session.sid])
+
+        # Less verbose error messages
+        command_args.extend(['--quiet'])
+        if not allow_js:
+            command_args.extend(['--disable-javascript'])
+
+        # Build paperformat args
+        command_args.extend(['--width', str(width)])
+        command_args.extend(['--height', str(height)])
+
+        return command_args
+
+    def _run_wkhtmltoimage(self, bodies, width, height):
+        command_args = self._build_wkhtmltoimage_args(width, height)
+        inputpath_arr = []
+        outputfd_arr = []
+        outputpath_arr = []
+        # this would be a great place for async :/
+        for i, body in enumerate(bodies):
+            infd, inputpath = tempfile.mkstemp(suffix='.html', prefix=f'report_image_html_input.tmp.{i}.')
+            with closing(os.fdopen(infd, 'wb')) as html_file:
+                html_file.write(body.encode())
+            inputpath_arr.append(inputpath)
+            outfd, outpath = tempfile.mkstemp(suffix='.png', prefix='report_image_output.tmp.')
+            outputfd_arr.append(outfd)
+            outputpath_arr.append(outpath)
+        output_image_arr = []
+        # start all processes THEN await
+        # TODO maybe put a limit on parallel processes...
+        process_arr = []
+        for input_path, output_path in zip(inputpath_arr, outputpath_arr):
+            wkhtmltoimage = [_get_wkhtmltoimage_bin()] + command_args + [input_path, output_path]
+            process_arr.append(subprocess.Popen(wkhtmltoimage, stdout=subprocess.PIPE, stderr=subprocess.PIPE))
+        # wait for processes in order they started (likely correlated to the order they finish)
```

Comment:
Not on the same thread, not even on the same process ^^

It starts up a new program, and the OS takes it from there. You have to opt into communication by using e.g. `communicate`, but otherwise it's actually completely unrelated to odoo once you call `Popen`

Reply:
Sorry, my question was wrong.

If I get your idea, you want to have a few processes running at the same time in order to take advantage of parallelization, but then we still communicate with them in order, so we have to wait for process one ending in order to communicate with process two and so on.

Then yes a small number of processes seems fine, or even one at a time if we want to simplify.